src/s390_crypto.c

```diff
@@ -675,6 +666,44 @@ int s390_get_functionlist(libica_func_list_element *pmech_list,
 		pmech_list[x].flags = 0;
 		pmech_list[x].property = 0;
 	}
+
+	/* Adjust the flags and properties for algorithms that are allowed in fips
+	 * mode, but not on any hardware, or not with any key length, curve etc. */
+	if (fips & ICA_FIPS_MODE) {
+		/* RSA >= 2048 bits in FIPS 140-3 mode */
+		switch (pmech_list[x].mech_mode_id) {
+		case RSA_KEY_GEN_ME:
+		case RSA_KEY_GEN_CRT:
+		case RSA_ME:
+		case RSA_CRT:
+			if (pmech_list[x].flags)
+				pmech_list[x].property = ICA_PROPERTY_RSA_FIPS;
+			else
+				pmech_list[x].property = 0;
+			break;
+		default:
+			break;
+		}
+
+		/* ECDSA/ECDH only via CPACF in FIPS 140-3 mode */
+		switch (pmech_list[x].mech_mode_id) {
+		case EC_DH:
+		case EC_DSA_SIGN:
+		case EC_DSA_VERIFY:
+		case EC_KGEN:
+			pmech_list[x].flags &= ~ICA_FLAG_DHW;
+			pmech_list[x].flags &= ~ICA_FLAG_SW;
+			if (pmech_list[x].flags) {
+				pmech_list[x].property &= ~ICA_PROPERTY_EC_BP;
+				pmech_list[x].property &= ~ICA_PROPERTY_EC_ED;
```

Comment:
Similar as above, are you actually restricting the API driven EC requests to not use BP and ED curves?

Reply:
See comment above, yes I think so, but will double-check.

Reply:
I don't see any fips checks (other than the general fips state is bad checks) that would prevent anyone from using certain curves when running in fips mode.... but maybe I am just not seeing it....
I also don't see any restriction on HW or SW usage when fips in on...... So even if OpenSSL would restrict certain curves when in fips mode, once it goes to the card or CPACF, then there is no restriction it seems.